exo/inference/pytorch/inference.py

```diff
@@ -0,0 +1,137 @@
+# experimental, based off of tinygrad/inference.py
+
+import numpy as np
+import torch
+import numpy as np
+import json
+from typing import Optional, Callable, Tuple
+from exo.inference.shard import Shard
+from exo.inference.inference_engine import InferenceEngine
+from exo.inference.pytorch.model.hf import ShardedHuggingFaceModel
+from exo.api.chatgpt_api import resolve_tokenizer
+from exo.helpers import DEBUG
+from transformers import DynamicCache
+
+class PyTorchDynamicShardInferenceEngine(InferenceEngine):
+    """
+    PyTorch Dynamic Shard Inference Engine for performing model inference with sharded models.
+    """
+
+    def __init__(self):
+        """
+        Initialize the inference engine.
+
+        Args:
+            debug (bool): If True, enables debug logging. Defaults to False.
+        """
+        self.shard = None
+        self.model = None
+        self.tokenizer = None
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```

Comment:
Are these the only options? I think supporting e.g. Mac with mps would be great since then you can run heterogeneous clusters.

One thing to try at some point would be mixing MLX and PyTorch and see if they are interoperable with exactly the same model.

Reply:
With pytorch I don't think mac is fully rolled out yet. There seems to be some work arounds but CUDA and CPU  are the only options on the pytorch download website. pytorch even stopped ROCm support for AMD

They have a nightly for testing MPS https://pytorch.org/blog/introducing-accelerated-pytorch-training-on-mac/

Reply:
What about this in the official "stable" docs: https://pytorch.org/docs/stable/notes/mps.html

Reply:
Will try that but currently no mac to test. When I get through these other fixes though I can definitely add it for you or other mac users to test.